expression/builtin_compare.go

```diff
@@ -1400,23 +1403,66 @@ func (c *compareFunctionClass) refineArgsByUnsignedFlag(ctx sessionctx.Context,
+				tp := col.RetType.Tp
+				upperBound := types.IntergerUnsignedUpperBound(tp)
 				if i == 1 {
 					op = symmetricOp[c.op]
 				}
 				if v == 0 && (op == opcode.LE || op == opcode.GT || op == opcode.NullEQ || op == opcode.EQ || op == opcode.NE) {
-					return args
+					// do nothing. In this situation, the args do not change.
+				} else if v <= 0 {
+					if colHasNotNullFlag { // Set the comparison result to true, need to consider the case of Null
+						// `unsigned_col < 0` equals to `1 < 0`,
+						// `unsigned_col > -1` equals to `1 > 0`,
+						// `unsigned_col <= -1` equals to `1 <= 0`,
+						// `unsigned_col >= 0` equals to `1 >= 0`,
+						// `unsigned_col == -1` equals to `1 == 0`,
+						// `unsigned_col != -1` equals to `1 != 0`,
+						// `unsigned_col <=> -1` equals to `1 <=> 0`,
+						// so we can replace the column argument with `1`, and the other constant argument with `0`.
+						args[i], args[1-i] = NewOne(), NewZero()
+					} else {
+						if op == opcode.LT || op == opcode.EQ { // Set the comparison result to false, need not to consider the case of Null
+							args[i], args[1-i] = NewOne(), NewZero()
+						}
+					}
+				} else if v > 0 && uint64(v) > upperBound {
+					if (op == opcode.NE || op == opcode.LE || op == opcode.LT) && colHasNotNullFlag {
+						args[i], args[1-i] = NewZero(), NewOne()
+					} else if op == opcode.EQ || op == opcode.GE || op == opcode.GT {
+						args[i], args[1-i] = NewZero(), NewOne()
+					}
+				}
+				return args
+			}
+			if !mysql.HasUnsignedFlag(con.RetType.Flag) && !mysql.HasUnsignedFlag(col.RetType.Flag) {
+				tp := col.RetType.Tp
+				lowerBound, upperBound := types.IntergerSignedLowerBound(tp), types.IntergerSignedUpperBound(tp)
+				lowerOverflow, upperOverflow := v < lowerBound, v > upperBound
+				op := c.op
+				if i == 1 {
+					op = symmetricOp[c.op]
+				}
+				// Check whether the range of the column value matches the range corresponding to the column type
+				// for example: column a is tinyint type -> [-128, 127]
+				// when there have a selection 'a > 200', the constant 200 is always greater than the maximum value that tinyint can achieve.
+				// So, the selection `a > 200` always be false.
+				// TODO: Consider the case where `v = upperbound || v = lowerBound`
+				if upperOverflow { // the constant value always greater than the column value
+					if op == opcode.EQ || op == opcode.GE || op == opcode.GT {
+						args[i], args[1-i] = NewZero(), NewOne()
+					} else if (op == opcode.LE || op == opcode.NE || op == opcode.LT) && colHasNotNullFlag {
+						args[i], args[1-i] = NewZero(), NewOne()
+					}
+				} else if lowerOverflow { // the constant value always less than the column value
+					if (op == opcode.EQ || op == opcode.GE || op == opcode.GT) && colHasNotNullFlag {
+						args[i], args[1-i] = NewOne(), NewZero()
+					} else if op == opcode.LE || op == opcode.NE || op == opcode.LT {
+						args[i], args[1-i] = NewOne(), NewZero()
+					}
```

Comment:
This function is so complicated / tricky now that it is hard to review, could you add more SQL examples in the comments to explain the purposes of these codes? or is it feasible to make it clean enough to separate this range comparison into another single function?